controllers/machinehealthcheck_controller.go

```diff
@@ -2,61 +2,504 @@ package controllers
+		if patchErr != nil {
+			log.Error(err, "failed to update status")
+		}
+		returnErr = utilerrors.NewAggregate([]error{patchErr, returnErr})
+	}()
+
+	// Return early if the object is paused
+	if utils.HasMHCPausedAnnotation(mhc) {
+		log.Info("Reconciliation is paused")
+		return result, nil
+	}
+
+	// For now we only support MHCs with RemediationTemplate set
+	if mhc.Spec.RemediationTemplate == nil {
+		log.Error(fmt.Errorf("empty RemediationTemplate"), "NodeHealthCheck only supports MachineHealthChecks with a RemediationTemplate")
+		// TODO event, status...?
+		return result, nil
+	}
+
+	// TODO add more checks like in NHC?!
+
+	log.Info("Reconciling")
+
+	// get targets
+	targets, err := resourceManager.GetMHCTargets(mhc)
+	if err != nil {
+		log.Error(err, "failed to get targets")
+		return result, err
+	}
+	totalTargets := len(targets)
+	metrics.ObserveMachineHealthCheckNodesCovered(mhc.Name, mhc.Namespace, totalTargets)
+
+	// health check all targets and reconcile mhc status
+	healthy, needRemediation, requeueIn, errList := r.checkHealth(targets)
+
+	healthyCount := 0
+	for _, healthyTarget := range healthy {
+		log.Info("handling healthy target", "target", healthyTarget.String())
+		_, err := resourceManager.HandleHealthyNode(healthyTarget.Node.GetName(), healthyTarget.Machine.GetName(), mhc)
+		if err != nil {
+			log.Error(err, "failed to handle healthy target", "target", healthyTarget.String())
+			return result, err
+		}
+		healthyCount++
+	}
+
+	mhc.Status.CurrentHealthy = &healthyCount
+	mhc.Status.ExpectedMachines = &totalTargets
+	unhealthyCount := totalTargets - healthyCount
+
+	// check if we didn't exceed max unhealthy nodes
+	maxUnhealthy, err := getMaxUnhealthy(mhc, log)
+	if err != nil {
+		log.Error(err, "failed to calculate max unhealthy allowed machines",
+			"maxUnhealthy", mhc.Spec.MaxUnhealthy, "total targets", totalTargets)
+		return result, err
+	}
+	mhc.Status.RemediationsAllowed = int32(getNrRemediationsAllowed(unhealthyCount, maxUnhealthy))
+
+	if !isRemediationsAllowed(unhealthyCount, maxUnhealthy) {
+
+		msg := fmt.Sprintf("Skipped remediation because the number of not started or unhealthy machines selected by the selector exceeds maxUnhealthy (total: %v, unhealthy: %v, maxUnhealthy: %v)",
+			totalTargets, unhealthyCount, mhc.Spec.MaxUnhealthy)
+		log.Info(msg)
+		r.Recorder.Event(mhc, corev1.EventTypeWarning, utils.EventReasonRemediationSkipped, msg)
+		metrics.ObserveMachineHealthCheckShortCircuitEnabled(mhc.Name, mhc.Namespace)
+
+		// Remediation not allowed, the number of not started or unhealthy machines exceeds maxUnhealthy
+		mhc.Status.RemediationsAllowed = 0
+		utils.SetMachineCondition(mhc, &v1beta1.Condition{
+			Type:     v1beta1.RemediationAllowedCondition,
+			Status:   corev1.ConditionFalse,
+			Severity: v1beta1.ConditionSeverityWarning,
+			Reason:   v1beta1.TooManyUnhealthyReason,
+			Message:  msg,
+		})
+
+		// no need to requeue if we don't have targets, or healthy nodes only
```

Comment:
I do not understand. Why should we not get here? :thinking:

Reply:
I meant that in the above if clause (`if !isRemediationsAllowed(unhealthyCount, maxUnhealthy) {`) we check if there are zero targets or zero unhealthy nodes (`if totalTargets == 0 || unhealthyCount == 0 {`), but I would expect this kind of check about the number of targets to happen earlier. Something like line 195, right after we find this value.

I am saying that because IIUC _GetMHCTargets_ will return X targets when there are X machines. What does MHC should do when there are no machines (= `totalTargets == 0`)? I would expect it to check that in line 195 before it continues with the reconcile loop.
Is that make sense?

Reply:
IMO we should not reach line 243 if `totalTargets == 0`

Reply:
but we would miss several status updates if we return earlier 🤷🏼‍♂️

Reply:
>miss several status updates

Status updates of _CurrentHealthy_, _ExpectedMachines_, and _RemediationsAllowed_? IIUC all of them are zero and they could be initialized right after we identify the value of _totalTargets_, in the rare case of not finding any machine.

`if totalTargets == 0 {
	zero := 0
	mhc.Status.CurrentHealthy = &zero
	mhc.Status.ExpectedMachines = &zero
	mhc.Status.RemediationsAllowed = 0
`

Reply:
Is that ok or am I missing something else?

Reply:
> if totalTargets == 0 { zero := 0 mhc.Status.CurrentHealthy = &zero mhc.Status.ExpectedMachines = &zero mhc.Status.RemediationsAllowed = 0 

Here we go, its not returning early only, but duplicating setting the status, for a rare edge case... why? 🤷🏼‍♂️ 
Also, this isn't new code, this is copied.

Reply:
A rare case indeed, and if it is copied then this is not the time to fix that.